ros1_foxglove_bridge/include/foxglove_bridge/msg_parser.hpp

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <map>
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace rospack {
+class Rosstackage;
+}
+
+namespace foxglove_bridge {
+
+struct MsgSpecException : public std::runtime_error {
+  using std::runtime_error::runtime_error;
+};
+struct MsgNotFoundException : public std::runtime_error {
+  using std::runtime_error::runtime_error;
+};
+struct InvalidMessageType : public std::runtime_error {
+  using std::runtime_error::runtime_error;
+};
+
+struct MsgSpec {
+  /// @brief Raw message definition.
+  std::string full_text;
+  /// @brief Field types.
+  std::vector<std::string> types;
+};
+
+class MsgParser {
+public:
+  explicit MsgParser(std::shared_ptr<rospack::Rosstackage> rospack = nullptr);
+
+  /// @brief Get the full message definition including definitions of all subtypes.
+  /// The output is equal to what `rosrun roslib gendeps -c <msg_file>` generates.
+  /// @param message_type Message type in the format <package_name>/<message_name>
+  /// @return String with the message schema and concatenated schemas of submessages.
+  std::string get_message_schema(const std::string& message_type);
+
+private:
+  std::string get_msg_file_path(const std::string& package_name, const std::string& message_name);
+  MsgSpec parse_msg_file(const std::string& package_name, const std::string& file_path);
+  std::shared_ptr<rospack::Rosstackage> rospack_;
+  std::map<std::string, MsgSpec> msg_spec_cache_;
```

Comment:
Does this need to be sorted alphabetically or can it use `std::unordered_map`?

Reply:
You are right, an unordered map is better here